File: 0496-next-greater-element-i/0496-next-greater-element-i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        
        
	int n = (int)nums1.size();
	map<int , int>mp;
	for(int i = 0 ; i < n ; i ++ ){
		mp[nums1[i]] = i; // val , idx
	}

	vector<int>ans(n , -1);
	stack<int> s;

	for(int i = 0 ; i < (int)nums2.size(); i ++){
		int cur_val = nums2[i];

		while(!s.empty() && cur_val > s.top()){
			int top_val = s.top();
			s.pop();
			int idx = mp[top_val];
			ans[idx] = cur_val;
		}

		if(mp.find(cur_val) != mp.end()){
			s.push(cur_val);
		}
	}

	return ans;
        
        
    }
};

File: 0496-next-greater-element-i/0496-next-greater-element-i.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        
        res = [-1] * len(nums1)
        nums1idx = {n: i for i , n in enumerate(nums1)}
        stack = []

        for i in range(len(nums2)):
            cur = nums2[i]
            while stack and cur > stack[-1]:
                val = stack.pop()
                idx = nums1idx[val]
                res[idx] = cur
            if cur in nums1idx:
                stack.append(cur)
        return res
